### DISCUSSION TITLE
Does library reproject EPSG 3857 into 4326 automatically

### DISCUSSION THREAD

**Discussion by EMaksymenko** [ngageoint/geopackage-mapcache-android#37] [Q&A] [answered]
I am trying to download GeoPackage from OpenTopoMap, OpenStreetMap, GoogleMaps and other online sources to test it with NASA WorldWind viewer, but when I select EPSG 4326 projection for this sources your app always process 1 tile per 5 sec and then it finish downloading with error message that no tiles were downloaded.

The same source URL with EPSG 3857 is downloading normally, but WorldWind does not support GeoPackage with this projection.

Example URL https://a.tile.opentopomap.org/{z}/{x}/{y}.png

Does application support automatic on the go reprojection of EPSG 3857 into 4326? If not, do you plan to implement it?

![Screenshot_20201113-181742_MapCache](https://user-images.githubusercontent.com/12487306/99094542-a0753200-25dc-11eb-943b-c24487f262b9.jpg)

**Comment by EMaksymenko** [comment]
Here is an example code to reproject downloaded tile if it was not implemented yet:
```
    public Bitmap transform3857to4326(Bitmap bitmap) {
        // Re-project mercator tile to equirectangular
        Bitmap trans = Bitmap.createBitmap(bitmap.getWidth(), bitmap.getHeight(), bitmap.getConfig());
        double miny = ((MercatorSector) sector).minLatPercent();
        double maxy = ((MercatorSector) sector).maxLatPercent();
        for (int y = 0; y < bitmap.getHeight(); y++) {
            double sy = 1.0 - y / (double) (bitmap.getHeight() - 1);
            double lat = sy * (sector.maxLatitude() - sector.minLatitude()) + sector.minLatitude();
            double dy = 1.0 - (MercatorSector.gudermannianInverse(lat) - miny) / (maxy - miny);
            dy = Math.max(0.0, Math.min(1.0, dy));
            int iy = (int) (dy * (bitmap.getHeight() - 1));
            for (int x = 0; x < bitmap.getWidth(); x++) {
                trans.setPixel(x, y, bitmap.getPixel(x, iy));
            }
        }
        return trans;
    }
```
Here is also a MercatorSector if required:
```
public class MercatorSector extends Sector {

    private final double minLatPercent, maxLatPercent;

    private MercatorSector(double minLatPercent, double maxLatPercent,
                           double minLongitude, double maxLongitude) {
        this.minLatPercent = minLatPercent;
        this.maxLatPercent = maxLatPercent;
        this.minLatitude = gudermannian(minLatPercent);
        this.maxLatitude = gudermannian(maxLatPercent);
        this.minLongitude = minLongitude;
        this.maxLongitude = maxLongitude;
    }

    public static MercatorSector fromDegrees(double minLatPercent, double maxLatPercent,
                                             double minLongitude, double maxLongitude) {
        return new MercatorSector(minLatPercent, maxLatPercent, minLongitude, maxLongitude);
    }

    static MercatorSector fromSector(Sector sector) {
        return new MercatorSector(gudermannianInverse(sector.minLatitude()),
                gudermannianInverse(sector.maxLatitude()),
                sector.minLongitude(), sector.maxLongitude());
    }

    static double gudermannianInverse(double latitude) {
        return Math.log(Math.tan(Math.PI / 4.0 + Math.toRadians(latitude) / 2.0)) / Math.PI;
    }

    private static double gudermannian(double percent) {
        return Math.toDegrees(Math.atan(Math.sinh(percent * Math.PI)));
    }

    double minLatPercent() {
        return minLatPercent;
    }

    double maxLatPercent()
    {
        return maxLatPercent;
    }

}
```

**Comment by bosborn** [comment]
The download tiles projection is meant to be set to the projection of a XYZ or WMS tile url.

EPSG 3857 for URLS like:
* https://a.tile.opentopomap.org/{z}/{x}/{y}.png [(single tile)](https://a.tile.opentopomap.org/5/6/12.png)
* https://osm.gs.mil/tiles/default/{z}/{x}/{y}.png [(single tile)](https://osm.gs.mil/tiles/default/5/6/12.png)
* http://nowcoast.noaa.gov/arcgis/rest/services/nowcoast/sat_meteo_imagery_time/MapServer/export?transparent=true&format=png&bbox={minLon},{minLat},{maxLon},{maxLat}&bboxSR=3857&imageSR=3857&size=256,256&f=image [(single tile)](https://nowcoast.noaa.gov/arcgis/rest/services/nowcoast/sat_meteo_imagery_time/MapServer/export?transparent=true&format=png&bbox=-112.500003,31.952164,-101.249996,40.979897&bboxSR=4326&imageSR=3857&size=256,256&f=image)

EPSG 4326 for something like:
* https://osm.gs.mil/tiles/default_pc/{z}/{x}/{y}.png [(single tile)](https://osm.gs.mil/tiles/default_pc/5/12/9.png)
* https://nowcoast.noaa.gov/arcgis/rest/services/nowcoast/sat_meteo_imagery_time/MapServer/export?transparent=true&format=png&bbox={minLon},{minLat},{maxLon},{maxLat}&bboxSR=4326&imageSR=4326&size=256,256&f=image [(single tile)](https://nowcoast.noaa.gov/arcgis/rest/services/nowcoast/sat_meteo_imagery_time/MapServer/export?transparent=true&format=png&bbox=-112.500003,31.952164,-101.249996,40.979897&bboxSR=4326&imageSR=4326&size=256,256&f=image)

For your problem (NASA WorldWind viewer unable to view EPSG 3857), your best option right now would be to find tile servers that support EPSG 4326.

Creating a GeoPackage with downloaded tiles and re-projecting those tiles on the fly is not reasonably feasible due to required overlapping pixels and tiles from the projections.  We are planning to add a full tile layer reprojection function to the libraries and apps which could be invoked upon tile download completion.

**Comment by EMaksymenko** [comment]
Thank you for quick answer.

What do you mean "required overlapping pixels"?

As I understand to reach the goal it is necessary only to make pixel by pixel transformation of each downloaded tile using the code example above and use different tile borders calculation in the viewer.

I am using this approach now for online sources in WorldWind (download and reproject tile on the go) and it looks ok without any artifacts.

I just search a possibility to make file cache and your library seems to be the best solution for this.

Do you think it is not useful feature to be able to use any popular online map source like OSM, OTM, Google, Yandex, Wiki in equirectangular projection?

Here is an example of reprojected tiles usage:
![Screenshot_20201114-112758_ComBat 4 Map](https://user-images.githubusercontent.com/12487306/99144192-7ec38b80-266c-11eb-975f-24e1554c17eb.jpg)

**Comment by bosborn** [comment]
It is certainly possible to project a single EPSG 3857 tile to EPSG 4326.  However, the projected tiles must be stored using the GeoPackage spec and tile storage requirements. The Tile Matrix Set defines the bounds (min x, min y, max x, max y) of the tiles in the projection.  For all tiles at each zoom level, a Tile Matrix defines the matrix width, matrix height, tile width, tile height, pixel x size, and pixel y size.  The pixel sizes must be constant throughout all tiles in the zoom level.

As an example, imagine receiving a 2 x 2 grid of 256 x 256 web mercator tiles, totaling 4 tiles.
WGS84 bounds: min longitude: -180, min latitude: 45.0, max longitude: -135.0, max latitude: 90.0

![topLeftWebMercator](https://user-images.githubusercontent.com/11667761/99316524-c3dcee80-2821-11eb-9e73-02f7dd5ec3ea.png) ![topRightWebMercator](https://user-images.githubusercontent.com/11667761/99316547-cc352980-2821-11eb-8791-474a993a38b1.png)
![bottomLeftWebMercator](https://user-images.githubusercontent.com/11667761/99316608-e3741700-2821-11eb-98d9-4d38357adffd.png) ![bottomRightWebMercator](https://user-images.githubusercontent.com/11667761/99316643-ee2eac00-2821-11eb-8f0f-38c169426495.png)

The pixel calculations in meters come out to
```
pixelXSize = 9783.9396205025768
pixelYSize = 28156.224329291363
```
These 4 tiles can each individually be projected to WGS84...

![topLeftWGS84OfWebMercator](https://user-images.githubusercontent.com/11667761/99316894-5bdad800-2822-11eb-862b-eab0d9f2cabc.png) ![topRightWGS84OfWebMercator](https://user-images.githubusercontent.com/11667761/99316909-61382280-2822-11eb-987d-827c2a30d2a7.png)
![bottomLeftWGS84OfWebMercator](https://user-images.githubusercontent.com/11667761/99316943-6e551180-2822-11eb-81ba-b3ead41ee951.png) ![bottomRightWGS84OfWebMercator](https://user-images.githubusercontent.com/11667761/99316955-744af280-2822-11eb-8a9e-79483611aa6c.png)

The pixel calculations in degrees for the top two tiles:
```
pixelXSize = 0.087890625
pixelYSize = 0.040202499408506398
```

The pixel calculations in degrees for the bottom two tiles:
```
pixelXSize = 0.087890625
pixelYSize = 0.1162472223876132
```

A single pixel x and y size must be stored for each zoom level of tiles.  Therefore, the WGS84 tiles must be projected and stored as (using "required overlapping pixels"):

![topLeftWGS84](https://user-images.githubusercontent.com/11667761/99317384-38fcf380-2823-11eb-9420-89b184aa08ed.png) ![topRightWGS84](https://user-images.githubusercontent.com/11667761/99317402-3f8b6b00-2823-11eb-9a5c-eb6d8fc296dd.png)
![bottomLeftWGS84](https://user-images.githubusercontent.com/11667761/99317496-65187480-2823-11eb-8dc1-941baabf3323.png) ![bottomRightWGS84](https://user-images.githubusercontent.com/11667761/99317505-6b0e5580-2823-11eb-8364-b4a81e1f0827.png)

The pixel calculations in degrees:
```
pixelXSize = 0.087890625
pixelYSize = 0.087890625
```

Tiles can be stored in any projection.  Our libraries and apps handle querying for tiles by XYZ, TMS, or bounding box in a desired projection.  The underlying tiles are queried, drawn to a single tile, and reprojected as needed.

The functionality we intend to add includes reprojecting any tile layer within the GeoPackage to a different projection (just not during on the fly downloading).

Our libraries also handle more than just the EPSG 3857 and EPSG 4326 projections.  All implementations are generic using a projection library.

> Do you think it is not useful feature to be able to use any popular online map source like OSM, OTM, Google, Yandex, Wiki in equirectangular projection?

Tiles from any source can currently be saved as a valid GeoPackage tile layer in their current projection.  Viewers / clients should implement the GeoPackage spec and reproject tiles as they need them.  Since you are using a viewer without this reprojection capability, you would be able to use the planned reprojection tile layer functionality to tune your GeoPackage.

**Comment by bosborn** [comment]
_Converted from an Issue to a Discussion_

**Comment by bosborn** [comment, marked as answer]
> The functionality we intend to add includes reprojecting any tile layer within the GeoPackage to a different projection (just not during on the fly downloading).

I have this functionality implemented in our iOS library and will port it to our Java/Android GeoPackage libraries next month.

**Reply by EMaksymenko** [reply]
Good news! I will look forward to this change.
It will allow to extract any popular online map source like Google or OSM into GeoPackage for WorldWindow using your application.

**Reply by EMaksymenko** [reply]
Hi, @bosborn 
Any news about this feature?

**Reply by bosborn** [reply]
I'm working on it in the core and Java libraries right now, followed by Android next.  Shortly after that, the libraries will all be released (iOS, Java, and Android).  It will then be useable from any app using the libraries.  Then it will need to get scheduled into the MapCache app to expose that functionality.

**Reply by bosborn** [reply]
Tile Reprojections have been released into the Android libraries
- [geopackage-android 5.0.0](https://github.com/ngageoint/geopackage-android/releases/tag/5.0.0)
- [geopackage-android-map 5.0.0](https://github.com/ngageoint/geopackage-android-map/releases/tag/5.0.0)

Support should be added to this app sometime down the road.

Updates were also released in the [Java library](https://github.com/ngageoint/geopackage-java/releases/tag/5.0.0).  If the reprojections needed are one time in advance operations, the library includes a few stand alone options.
- [Tile Reprojections Standalone Utility](https://github.com/ngageoint/geopackage-java#tile-reprojections)
- [SQLite Exec Standalone Utility](https://github.com/ngageoint/geopackage-java/tree/master/script/sqlite-exec#sqlite-exec)

**Reply by emartynov** [reply]
Hi @bosborn, I'm also working on a similar task and I'm looking to use your library. I see that `TileGenerator` is using projections to recalculate boundaries, however, later it just calculates and loads tiles without any reprojections.

Do, I understand correctly, that I have to use a standalone step later to run reprojections on the tiles in the file?

**Reply by emartynov** [reply]
One more question - it looks like the tile generator doesn't calculate the tile coordinates for non-mercator projections. I had issues that cache was drawing some different parts of map until I changed the base URL from `https://osm.gs.mil/tiles/default/<zoom>/<x>/<y>.png` to `https://osm.gs.mil/tiles/default_pc/<zoom>/<x>/<y>.png`

**Reply by bosborn** [reply]
[TileGenerator](https://github.com/ngageoint/geopackage-android/blob/master/geopackage-sdk/src/main/java/mil/nga/geopackage/tiles/TileGenerator.java) is used to create and store a set of tiles within the GeoPackage, such as from a URL ([UrlTileGenerator](https://github.com/ngageoint/geopackage-android/blob/master/geopackage-sdk/src/main/java/mil/nga/geopackage/tiles/UrlTileGenerator.java)) or existing features ([FeatureTileGenerator](https://github.com/ngageoint/geopackage-android/blob/master/geopackage-sdk/src/main/java/mil/nga/geopackage/tiles/features/FeatureTileGenerator.java)).  The Tile Reprojections mentioned above are for one time changing the projection of the physically stored tiles in the GeoPackage (such as once after downloading).  On the fly reprojections for display are handled through [GeoPackageTileRetriever](https://github.com/ngageoint/geopackage-android/blob/master/geopackage-sdk/src/main/java/mil/nga/geopackage/tiles/retriever/GeoPackageTileRetriever.java) using [TileCreator](https://github.com/ngageoint/geopackage-android/blob/master/geopackage-sdk/src/main/java/mil/nga/geopackage/tiles/retriever/TileCreator.java) and the [BoundedOverlay](https://github.com/ngageoint/geopackage-android-map/blob/master/geopackage-map/src/main/java/mil/nga/geopackage/map/tiles/overlay/BoundedOverlay.java) `TileProvider` retrieved from [GeoPackageOverlayFactory](https://github.com/ngageoint/geopackage-android-map/blob/master/geopackage-map/src/main/java/mil/nga/geopackage/map/tiles/overlay/GeoPackageOverlayFactory.java).  You do not need to use a standalone step later if you are satisfied with the projection the tiles are stored as.  Display as any other projection is supported on demand.

The `UrlTileGenerator` projection used should match the projection of the tiles being downloaded.  The `default_pc` (Platte Carre) OSM url is for downloading WGS84 (EPSG:4326) tiles.  WGS84 is supported by the TileGenerator.  The `default` OSM URL provides Web Mercator tiles (EPSG:3857).  Match your specified `TileGenerator` projection with the URL you are using.
